### PR TITLE
Populates ReservationWindow based on # of open reservations

### LINKED ISSUE #122: Return wait time using the number of people ahead of a caller, and a constant delay between callers
Using the number of people ahead of a caller in #121 and a passed in delay between each caller being handled (to start, this delay should be related to the number of calls handled in an hour, which is a parameter to the consumer script), return the wait time.
 - The wait time calculation is currently in a hacked place, on a getter that calculates the value on the fly
 - Might be better ways to have dynamic value in Mongo REST (for example, using a projection or the @Value annotation)

### PR DESCRIPTION
**Pull request by kebrahim**
Fixes #122 

I tried a few different ways to do this. Overriding default behavior for ReservationRepository seemed to make the most sense, but LMK if you have concerns.

- [X] Tests pass
- [X] Appropriate changes to README are included in PR
- [ ] Feature has been fleshed out in design document

### CODE REVIEW

File: callmeback/src/main/java/org/google/callmeback/api/CustomizedReservationRepository.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+import org.springframework.data.mongodb.core.query.Query;
+
+/** A ReservationRepository that overrides various methods of MongoRepository. */
+public interface CustomizedReservationRepository<T, ID> {
+  /**
+   * Returns a Reservation by the specified ID, including populating the ReservationWindow, based
+   * on the number of reservations in the system.
+   */
+  Optional<T> findById(ID id);
+
```
Comment (mscharfstein) [inline callmeback/src/main/java/org/google/callmeback/api/CustomizedReservationRepository.java:19]: what is T? and S below? do these names have to be so minimalist?

File: callmeback/src/main/java/org/google/callmeback/api/CustomizedReservationRepository.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+  private static final int AVERAGE_CALL_DURATION_MINUTES = 10;
+  private static final int WINDOW_LENGTH_MINUTES = 30;
+
+  @Override
+  public Optional<T> findById(ID id) {
+    Reservation reservation = mongoTemplate.findById(id, Reservation.class);
+    reservation.window = getWindow(reservation.reservationCreatedDate);
+    return Optional.of((T)reservation);
+  }
```
Comment (mscharfstein) [inline callmeback/src/main/java/org/google/callmeback/api/CustomizedReservationRepository.java:40]: does this turn T into a "reservation"?

File: callmeback/src/main/java/org/google/callmeback/api/CustomizedReservationRepository.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+  /**
+   * Returns a ReservationWindow, incorporating the number of reservations created prior to the
+   * specified date, which do not currently have any reservation events associated.
+   */
+  private ReservationWindow getWindow(Date requestDate) {
+    ReservationWindow window = new ReservationWindow();
+    Query query = new Query(
+        Criteria.where("events").is(null).and("reservationCreatedDate").lt(requestDate));
+    long countReservations = mongoTemplate.count(query, Reservation.class);
+    long expectedWaitTimeMins = countReservations * AVERAGE_CALL_DURATION_MINUTES;
```
Comment (mscharfstein) [inline callmeback/src/main/java/org/google/callmeback/api/CustomizedReservationRepository.java:59]: I don't think the wait time should be countReservations * call duration, but instead countReservations * queueHandleDuration - not sure what to call it, but it should be the amount of time it takes to get the next person off of the queue. That would only be the same as the call duration if there was a single agent and the agent never took breaks. Our call center representation can model multiple agents such that callers are taken off of the queue faster (potentially) than an individual call takes

File: callmeback/src/main/java/org/google/callmeback/api/CustomizedReservationRepository.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+    Date current = new Date();
+    window.min = Date.from(current.toInstant().plus(Duration.ofMinutes(expectedWaitTimeMins)));
```
Comment (mscharfstein) [inline callmeback/src/main/java/org/google/callmeback/api/CustomizedReservationRepository.java:62]: don't feel strongly but should this be the expected and min and max are around this one?

File: callmeback/src/main/java/org/google/callmeback/api/ReservationRepository.java [hunk 5]
```diff
@@ -8,7 +8,10 @@
 
 @RepositoryRestResource
 public interface ReservationRepository
-    extends MongoRepository<Reservation, String>, QueueingRepository {
+    extends MongoRepository<Reservation, String>,
+        QueueingRepository,
+        CustomizedReservationRepository<Reservation, String> {
+
   /** Returns a list of Reservations with the specified topic. */
   List<Reservation> findByTopic(@Param("topic") String topic);
```
Comment (mscharfstein) [inline callmeback/src/main/java/org/google/callmeback/api/ReservationRepository.java:17]: do we still need countByEventsNullAndReservationCreatedDateLessThan below then?

File: callmeback/src/main/java/org/google/callmeback/api/CustomizedReservationRepository.java (hunk 1 above)
Comment (kebrahim) [reply callmeback/src/main/java/org/google/callmeback/api/CustomizedReservationRepository.java:19]: these are the generic types defined in CrudRepository - https://docs.spring.io/spring-data/commons/docs/current/api/org/springframework/data/repository/CrudRepository.html

this was the recommended way to override the default methods, at least from this article - https://medium.com/clarityai-engineering/overriding-default-methods-of-spring-mongorepository-e7ca6a637132

anyway, i added additional documentation to explain what they refer to

File: callmeback/src/main/java/org/google/callmeback/api/CustomizedReservationRepository.java (hunk 2 above)
Comment (kebrahim) [reply callmeback/src/main/java/org/google/callmeback/api/CustomizedReservationRepository.java:40]: yeah, T is Reservation (see comment at top of the interface)

File: callmeback/src/main/java/org/google/callmeback/api/CustomizedReservationRepository.java (hunk 3 above)
Comment (kebrahim) [reply callmeback/src/main/java/org/google/callmeback/api/CustomizedReservationRepository.java:59]: done

File: callmeback/src/main/java/org/google/callmeback/api/CustomizedReservationRepository.java (hunk 4 above)
Comment (mscharfstein) [reply callmeback/src/main/java/org/google/callmeback/api/CustomizedReservationRepository.java:62]: sgtm

File: callmeback/src/main/java/org/google/callmeback/api/CustomizedReservationRepository.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+  private MongoTemplate mongoTemplate;
+
+  private static final int AVERAGE_CALL_DURATION_MINUTES = 10;
+  private static final int WINDOW_LENGTH_MINUTES = 30;
+
+  @Override
+  public Optional<T> findById(ID id) {
+    Reservation reservation = mongoTemplate.findById(id, Reservation.class);
+    reservation.window = getWindow(reservation.reservationCreatedDate);
+    return Optional.of((T)reservation);
```
Comment (egracer) [inline callmeback/src/main/java/org/google/callmeback/api/CustomizedReservationRepository.java:39]: If 'T' is always going to be a reservation, why not just declare it explicitly as the type? Why do we need this to remain generic?

File: callmeback/src/main/java/org/google/callmeback/api/ReservationRepository.java (hunk 5 above)
Comment (kebrahim) [reply callmeback/src/main/java/org/google/callmeback/api/ReservationRepository.java:17]: you're right, we don't. removed.

File: callmeback/src/main/java/org/google/callmeback/api/CustomizedReservationRepository.java (hunk 6 above)
Comment (kebrahim) [reply callmeback/src/main/java/org/google/callmeback/api/CustomizedReservationRepository.java:39]: apparently, you need to keep the method signatures the same or Spring won't correctly the overridden methods - see the "Overriding MongoRepository" section here - https://medium.com/clarityai-engineering/overriding-default-methods-of-spring-mongorepository-e7ca6a637132

Comment (egracer) [reply callmeback/src/main/java/org/google/callmeback/api/CustomizedReservationRepository.java:49]: Interesting, and bizarre! But automagical :)

File: callmeback/src/main/java/org/google/callmeback/api/CustomizedReservationRepository.java [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+}
+
+class CustomizedReservationRepositoryImpl<T, ID> implements CustomizedReservationRepository<T, ID> {
+
+  @Autowired
+  private MongoTemplate mongoTemplate;
+
+  // (Hard-coded) Average time it takes for a call to be taken off the queue, after a reservation is
+  // made
+  private static final int AVERAGE_CALL_HANDLING_MINS = 10;
```
Comment (mscharfstein) [inline callmeback/src/main/java/org/google/callmeback/api/CustomizedReservationRepository.java:40]: maybe its AVERAGE_WAIT_TIME_MINS ? its really the wait time

Comment (kebrahim) [reply callmeback/src/main/java/org/google/callmeback/api/CustomizedReservationRepository.java]: done

**Pull request merged**